platforms/linux/arm.toolchain.cmake

```diff
@@ -48,32 +48,40 @@ if(NOT DEFINED ARM_LINUX_SYSROOT AND DEFINED GNU_MACHINE)
   set(ARM_LINUX_SYSROOT /usr/${GNU_MACHINE}${FLOAT_ABI_SUFFIX})
 endif()
 
-if(NOT DEFINED CMAKE_CXX_FLAGS)
-  set(CMAKE_CXX_FLAGS           "" CACHE INTERNAL "")
-  set(CMAKE_C_FLAGS             "" CACHE INTERNAL "")
-  set(CMAKE_SHARED_LINKER_FLAGS "" CACHE INTERNAL "")
-  set(CMAKE_MODULE_LINKER_FLAGS "" CACHE INTERNAL "")
-  set(CMAKE_EXE_LINKER_FLAGS    "" CACHE INTERNAL "")
-
-  set(CMAKE_CXX_FLAGS           "${CMAKE_CXX_FLAGS} -fdata-sections -Wa,--noexecstack -fsigned-char -Wno-psabi")
-  set(CMAKE_C_FLAGS             "${CMAKE_C_FLAGS} -fdata-sections -Wa,--noexecstack -fsigned-char -Wno-psabi")
-  if(CMAKE_SYSTEM_PROCESSOR STREQUAL arm)
-    set(CMAKE_CXX_FLAGS           "-mthumb ${CMAKE_CXX_FLAGS}")
-    set(CMAKE_C_FLAGS             "-mthumb ${CMAKE_C_FLAGS}")
-    set(CMAKE_EXE_LINKER_FLAGS    "${CMAKE_EXE_LINKER_FLAGS} -Wl,-z,nocopyreloc")
-  endif()
-  if(CMAKE_SYSTEM_PROCESSOR STREQUAL arm)
-    set(ARM_LINKER_FLAGS "-Wl,--fix-cortex-a8 -Wl,--no-undefined -Wl,--gc-sections -Wl,-z,noexecstack -Wl,-z,relro -Wl,-z,now")
-  elseif(CMAKE_SYSTEM_PROCESSOR STREQUAL aarch64)
-    set(ARM_LINKER_FLAGS "-Wl,--no-undefined -Wl,--gc-sections -Wl,-z,noexecstack -Wl,-z,relro -Wl,-z,now")
+# == Compiler flags
+set(common_cc_opt "-fdata-sections -Wa,--noexecstack -fsigned-char -Wno-psabi")
```

Comment:
Looks like these options are lost: https://pullrequest.opencv.org/buildbot/builders/precommit_armv8/builds/100061

Reply:
This is because cmake version on buildbot is too old (3.5.1) for this [feature](https://cmake.org/cmake/help/latest/variable/CMAKE_LANG_FLAGS_INIT.html) (`CMAKE_<LANG>_FLAGS_INIT`). I propose raising minimum cmake version to 3.7.

Reply:
My experiments has shown that unfortunately `FLAGS_INIT` vars do not work with regular `FLAGS` variables as I expected. They are not appended to initial value, but override it:

```
cmake \
  -DCMAKE_TOOLCHAIN_FILE=/work/opencv/platforms/linux/riscv64-clang.toolchain.cmake \
  ../opencv
# result: -march=rv64gc ...
# this is flag from our toolchain file

cmake \
  -DCMAKE_TOOLCHAIN_FILE=/work/opencv/platforms/linux/riscv64-clang.toolchain.cmake \
  -DCMAKE_CXX_FLAGS="-Werror" \
  ../opencv
# result: -Werror ...
# no -march flags from toolchain file
```
Correct way to add custom flags to cross-compilation in this case will be via
* `CMAKE_<LANG>_FLAGS_<CONFIG>` variable, or
* `<LANG>FLAGS` environment variable, or
* OpenCV-specific variables `OPENCV_EXTRA_<LANG>_FLAGS`, or
* custom toolchain-file.

I believe this behavior is acceptable because this is how cmake works.